### PR TITLE
Prevent submitting papers in some circumstances

### PR DESCRIPTION
**Pull request by mic4ael**

### CODE REVIEW

File: indico/modules/events/papers/controllers/display.py [hunk 1]
```diff
@@ -43,6 +43,12 @@ class RHSubmitPaper(RHPaperBase):
     PAPER_REQUIRED = False
     ALLOW_LOCKED = True
 
+    def _check_paper_protection(self):
+        if not RHPaperBase._check_paper_protection(self):
+            return False
+        last_revision = self.contribution._paper_last_revision
+        return last_revision is None or last_revision.state == PaperRevisionState.to_be_corrected
```
Comment (ThiefMaster) [inline indico/modules/events/papers/controllers/display.py]: Do we also want to require the last revision to have been submitted by the same user? Or should any author be able to submit a new one once it's in "to be corrected"?

File: indico/modules/events/papers/util.py [hunk 2]
```diff
@@ -85,5 +85,9 @@ def has_contributions_with_user_paper_submission_rights(event, user):
     return _query_contributions_with_user_paper_submission_rights(event, user).has_rows()
 
 
-def get_contributions_with_user_paper_submission_rights(event, user):
-    return _query_contributions_with_user_paper_submission_rights(event, user).all()
+def get_contributions_with_submission_rights_and_proper_state(event, user):
```
Comment (ThiefMaster) [inline indico/modules/events/papers/util.py]: Can't we find a nicer name for this? `get_user_submittable_contributions` isn't super nice either but doesn't sound as bad as both the previous name or the updated one.

File: indico/modules/events/papers/controllers/display.py (hunk 1 above)
Comment (mic4ael) [reply indico/modules/events/papers/controllers/display.py]: @pferreir what do you think?

Comment (mic4ael) [reply indico/modules/events/papers/controllers/display.py]: But i guess we shouldn't limit it to the same user unless there is a good reason for that.

Comment (pferreir) [reply indico/modules/events/papers/controllers/display.py]: I think we should not limit it to the same person. Being flexible in this case is probably the best.

Comment (pferreir) [reply indico/modules/events/papers/controllers/display.py]: Think of co-authors working collaboratively on a paper.

File: indico/modules/events/papers/util.py [hunk 3]
```diff
@@ -85,5 +85,9 @@ def has_contributions_with_user_paper_submission_rights(event, user):
     return _query_contributions_with_user_paper_submission_rights(event, user).has_rows()
 
 
-def get_contributions_with_user_paper_submission_rights(event, user):
-    return _query_contributions_with_user_paper_submission_rights(event, user).all()
+def get_user_submittable_contributions(event, user):
+    allowed_states = [PaperRevisionState.to_be_corrected]
+    return (_query_contributions_with_user_paper_submission_rights(event, user)
+            .filter(db.or_(Contribution._paper_last_revision == None,  # noqa
```
Comment (pferreir) [inline indico/modules/events/papers/util.py]: why `noqa`?

Comment (ThiefMaster) [reply indico/modules/events/papers/util.py]: `== None` (sqlalchemy does not implement `is_` and `isnot` there)

Comment (pferreir) [reply indico/modules/events/papers/util.py]: Ah right.

**Comment by ThiefMaster** [comment]
Merged in e2f4545ef (and fixed in 1c86d550e)

**Pull request closed without merging**